File: java/array_search/SentinelLinearSearch.java
public class SentinelLinearSearch {
    public static int sentinelLinearSearch(int[] array, int target) {
        if (array.length == 0) return -1;

        int lastIndex = array.length - 1;
        int savedLast = array[lastIndex];
        array[lastIndex] = target;

        int index = 0;
        while (array[index] != target) {
            index++;
        }

        array[lastIndex] = savedLast;

        if (index < lastIndex || savedLast == target) {
            return index;
        }

        return -1;
    }
}

File: python/array_search/sentinel_linear_search.py
def sentinel_linear_search(array, target):
    if not array:
        return -1 

    last_index = len(array) - 1
    saved_last = array[last_index]
    array[last_index] = target

    index = 0 
    while array[index] != target:
        index += 1

    array[last_index] = saved_last

    if index < last_index or saved_last == target: 
        return index 

    return -1 
